### PR TITLE
Fix #107 update gosc log checking conditions

### LINKED ISSUE #22: [v1.2][New Feature][Linux] Collect cloud-init logs after GOSC
##### SUMMARY

Run `cloud-init collect-logs` can collect more information, for example
```
root@gosc-dhcp-vm-01 [ ~ ]# cloud-init collect-logs 
Wrote /root/cloud-init.tar.gz
root@gosc-dhcp-vm-01 [ ~ ]# tar -zxvf cloud-init.tar.gz 
cloud-init-logs-2021-04-20/
cloud-init-logs-2021-04-20/run/
cloud-init-logs-2021-04-20/run/cloud-init/
cloud-init-logs-2021-04-20/run/cloud-init/cloud-init-generator.log
cloud-init-logs-2021-04-20/run/cloud-init/ds-identify.log
cloud-init-logs-2021-04-20/run/cloud-init/cloud.cfg
cloud-init-logs-2021-04-20/run/cloud-init/.ds-identify.result
cloud-init-logs-2021-04-20/run/cloud-init/enabled
cloud-init-logs-2021-04-20/run/cloud-init/instance-data.json
cloud-init-logs-2021-04-20/run/cloud-init/instance-data-sensitive.json
cloud-init-logs-2021-04-20/run/cloud-init/.instance-id
cloud-init-logs-2021-04-20/run/cloud-init/network-config-ready
cloud-init-logs-2021-04-20/run/cloud-init/status.json
cloud-init-logs-2021-04-20/run/cloud-init/result.json
cloud-init-logs-2021-04-20/cloud-init-output.log
cloud-init-logs-2021-04-20/cloud-init.log
cloud-init-logs-2021-04-20/journal.txt
cloud-init-logs-2021-04-20/dmesg.txt
cloud-init-logs-2021-04-20/dpkg-version
cloud-init-logs-2021-04-20/version
```
We can add it into cloud-init GOSC test cases, and collect and retrieve these information after GOSC for furhter analysis or bug report.

##### ISSUE TYPE
- Feature Idea

##### ENVIRONMENT / OS

Linux

### LINKED ISSUE #107: [Bug] [Linux] Linux toolsDeployPkg log has been updated for successful and failed GOSC
##### SUMMARY

The condition for checking gosc_log_success has been changed, which need to be updated.  See pr 2333071 and 2636532.
```
2021-07-21 12:26:11,021 | TASK [Assert perl guest customization checks all PASS] *****
task path: /root/ansible-vsphere-gos-validation/linux/guest_customization/linux_gosc_verify.yml:54
failed: [localhost] => (item={'key': 'gosc_log_success', 'value': False}) => {
    "ansible_loop_var": "item",
    "assertion": false,
    "changed": false,
    "evaluated_to": false,
    "item": {
        "key": "gosc_log_success",
        "value": false
    },
    "msg": "perl guest customization with dhcp IP failed on: {'key': 'gosc_log_success', 'value': False}"
}
```

### PR DESCRIPTION
**Pull request by keirazhang**
Fix #107 Even though 'telinit returned error 127' message is shown in toolsDeployPkg.log, it doesn't mean that guest OS was not rebooted after GOSC. By checking with GOSC team, when 'Ran DeployPkg_DeployPackageFromFile successfully' is shown in toolsDeployPkg.log, it means guest customization process finished, we can check all customized properties like hostname, DNS, ip, etc. As long as these guest properties are customized successfully, GOSC tests pass.

Fix #22 cloud-init logs are collected into cloud-init.tar.gz

### CODE REVIEW

**Comment by keirazhang** [comment]
Perl GOSC failed with Photon OS 3.0, need to double check root cause.

File: docs/KNOWN_ISSUES.md [hunk 1]
```diff
@@ -20,20 +20,6 @@ guest_customization/gosc_cloudinit_staticip.yml
 * Workaround: Please wait for the fix, or make sure not deleting VMs while the test is running.
 * Affected test cases: Testing will fail if any of the task failed due to this issue.
 
-## VMware tools
-1. open-vm-tools 11.1.0 and later
-* Failure: Error logs in 'var/log/vmware-imc/toolsDeployPkg.log' of Perl guest OS customization in RHEL/CentOS/Oracle Linux 7.x and 8.x, e.g.,
-```
-[2021-03-18T15:52:10.974Z] [   error] Customization command failed with exitcode: 127, stderr: ''.
-[2021-03-18T15:52:11.974Z] [   error] Failed to reboot, telinit returned error 127.
```
Comment (Tomorrow9) [inline docs/KNOWN_ISSUES.md:28]: This "failed to reboot" error message is not an error actually?

Comment (keirazhang) [reply docs/KNOWN_ISSUES.md:28]: Even though we see this message in toolsDeployPkg.log, it doesn't mean that the guest was not rebooted. By checking with Pengpeng, he said that as long as all customized properties are successfully set in guest. We can consider that GOSC succeeds. At present, if 'uncaught signal 15' was reported in toolsDeployPkg.log, it is an real error(see PR 2716292). Maybe I can add checking 'uncaught signal 15' and record this as a known issue.

**Comment by keirazhang** [comment]
```
VM information:
+--------------------------------------------------------------------------------------+
| Name                      | test_vm                                                  |
+--------------------------------------------------------------------------------------+
| IP                        | x.x.x.x                                                  |
+--------------------------------------------------------------------------------------+
| Guest OS Type             | RedHat 7.9 x86_64                                        |
+--------------------------------------------------------------------------------------+
| VMTools Version           | 11.0.5.17716 (build-15389592)                            |
+--------------------------------------------------------------------------------------+
| CloudInit Version         | 19.4                                                     |
+--------------------------------------------------------------------------------------+
| Config Guest Id           | rhel7_64Guest                                            |
+--------------------------------------------------------------------------------------+
| Hardware Version          | vmx-19                                                   |
+--------------------------------------------------------------------------------------+
| GuestInfo Guest Id        | rhel7_64Guest                                            |
+--------------------------------------------------------------------------------------+
| GuestInfo Guest Full Name | Red Hat Enterprise Linux 7 (64-bit)                      |
+--------------------------------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                                               |
+--------------------------------------------------------------------------------------+
| GuestInfo Detailed Data   | bitness='64'                                             |
|                           | distroName='Red Hat Enterprise Linux Server'             |
|                           | distroVersion='7.9'                                      |
|                           | familyName='Linux'                                       |
|                           | kernelVersion='3.10.0-1160.el7.x86_64'                   |
|                           | prettyName='Red Hat Enterprise Linux Server 7.9 (Maipo)' |
+--------------------------------------------------------------------------------------+


Test Results (Total: 4, Failed: 0, No Run: 0, Elapsed Time: 00:27:17):
+----------------------------------------------+
| Name                    | Status | Exec Time |
+----------------------------------------------+
| gosc_perl_dhcp          | Passed | 00:05:30  |
| gosc_perl_staticip      | Passed | 00:04:31  |
| gosc_cloudinit_dhcp     | Passed | 00:07:40  |
| gosc_cloudinit_staticip | Passed | 00:07:24  |
+----------------------------------------------+
```
```
VM information:
+-------------------------------------------------------------------------------+
| Name                      | test_vm                                           |
+-------------------------------------------------------------------------------+
| IP                        | x.x.x.x                                           |
+-------------------------------------------------------------------------------+
| Guest OS Type             | RedHat 8.4 x86_64                                 |
+-------------------------------------------------------------------------------+
| VMTools Version           | 11.2.0.23855 (build-16938113)                     |
+-------------------------------------------------------------------------------+
| CloudInit Version         | 20.3                                              |
+-------------------------------------------------------------------------------+
| Config Guest Id           | rhel8_64Guest                                     |
+-------------------------------------------------------------------------------+
| Hardware Version          | vmx-19                                            |
+-------------------------------------------------------------------------------+
| GuestInfo Guest Id        | rhel8_64Guest                                     |
+-------------------------------------------------------------------------------+
| GuestInfo Guest Full Name | Red Hat Enterprise Linux 8 (64-bit)               |
+-------------------------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                                        |
+-------------------------------------------------------------------------------+
| GuestInfo Detailed Data   | bitness='64'                                      |
|                           | distroName='Red Hat Enterprise Linux'             |
|                           | distroVersion='8.4'                               |
|                           | familyName='Linux'                                |
|                           | kernelVersion='4.18.0-305.el8.x86_64'             |
|                           | prettyName='Red Hat Enterprise Linux 8.4 (Ootpa)' |
+-------------------------------------------------------------------------------+


Test Results (Total: 4, Failed: 0, No Run: 0, Elapsed Time: 00:24:22):
+----------------------------------------------+
| Name                    | Status | Exec Time |
+----------------------------------------------+
| gosc_perl_dhcp          | Passed | 00:05:56  |
| gosc_perl_staticip      | Passed | 00:04:57  |
| gosc_cloudinit_dhcp     | Passed | 00:05:41  |
| gosc_cloudinit_staticip | Passed | 00:05:18  |
+----------------------------------------------+
```

```

VM information:
+---------------------------------------------------------------+
| Name                      | test_vm                           |
+---------------------------------------------------------------+
| IP                         | x.x.x.x                          |
+---------------------------------------------------------------+
| Guest OS Type             | VMware Photon OS 4.0 x86_64       |
+---------------------------------------------------------------+
| VMTools Version           | 11.2.5.26209 (build-17337674)     |
+---------------------------------------------------------------+
| CloudInit Version         | 21.3                              |
+---------------------------------------------------------------+
| Config Guest Id           | vmwarePhoton64Guest               |
+---------------------------------------------------------------+
| Hardware Version          | vmx-13                            |
+---------------------------------------------------------------+
| GuestInfo Guest Id        | vmwarePhoton64Guest               |
+---------------------------------------------------------------+
| GuestInfo Guest Full Name | VMware Photon OS (64-bit)         |
+---------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                        |
+---------------------------------------------------------------+
| GuestInfo Detailed Data   | bitness='64'                      |
|                           | distroName='VMware Photon OS'     |
|                           | distroVersion='4.0'               |
|                           | familyName='Linux'                |
|                           | kernelVersion='5.10.4-13.ph4-esx' |
|                           | prettyName='VMware Photon OS 4.0' |
+---------------------------------------------------------------+


Test Results (Total: 4, Failed: 0, No Run: 0, Elapsed Time: 00:21:33):
+----------------------------------------------+
| Name                    | Status | Exec Time |
+----------------------------------------------+
| gosc_perl_dhcp          | Passed | 00:05:52  |
| gosc_perl_staticip      | Passed | 00:04:12  |
| gosc_cloudinit_dhcp     | Passed | 00:04:53  |
| gosc_cloudinit_staticip | Passed | 00:04:34  |
+----------------------------------------------+
```

```
VM information:
+---------------------------------------------------------------+
| Name                      | ubuntu-21.04                      |
+---------------------------------------------------------------+
| IP                        | x.x.x.x                           |
+---------------------------------------------------------------+
| Guest OS Type             | Ubuntu 21.04 x86_64               |
+---------------------------------------------------------------+
| VMTools Version           | 11.2.5.26209 (build-17337674)     |
+---------------------------------------------------------------+
| CloudInit Version         | 21.2-3-g899bfaa9-0ubuntu2~21.04.1 |
+---------------------------------------------------------------+
| Config Guest Id           | ubuntu64Guest                     |
+---------------------------------------------------------------+
| Hardware Version          | vmx-10                            |
+---------------------------------------------------------------+
| GuestInfo Guest Id        | ubuntu64Guest                     |
+---------------------------------------------------------------+
| GuestInfo Guest Full Name | Ubuntu Linux (64-bit)             |
+---------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                        |
+---------------------------------------------------------------+
| GuestInfo Detailed Data   | bitness='64'                      |
|                           | distroName='Ubuntu'               |
|                           | distroVersion='21.04'             |
|                           | familyName='Linux'                |
|                           | kernelVersion='5.11.0-22-generic' |
|                           | prettyName='Ubuntu 21.04'         |
+---------------------------------------------------------------+


Test Results (Total: 4, Failed: 1, No Run: 0, Elapsed Time: 00:32:04):
+------------------------------------------------+
| Name                    |   Status | Exec Time |
+------------------------------------------------+
| gosc_perl_dhcp          |   Passed | 00:10:51  |
| gosc_perl_staticip      |   Passed | 00:09:04  |
| gosc_cloudinit_dhcp     | * Failed | 00:05:11  |
| gosc_cloudinit_staticip |   Passed | 00:04:46  |
+------------------------------------------------+
```

**Comment by keirazhang** [comment]
```

VM information:
+---------------------------------------------------------------------------+
| Name                      | almalinux-8.4                                 |
+---------------------------------------------------------------------------+
| IP                        | 10.185.229.26                                 |
+---------------------------------------------------------------------------+
| Guest OS Type             | AlmaLinux 8.4 x86_64                          |
+---------------------------------------------------------------------------+
| VMTools Version           | 11.2.0.23855 (build-16938113)                 |
+---------------------------------------------------------------------------+
| CloudInit Version         | 20.3                                          |
+---------------------------------------------------------------------------+
| Config Guest Id           | other4xLinux64Guest                           |
+---------------------------------------------------------------------------+
| Hardware Version          | vmx-19                                        |
+---------------------------------------------------------------------------+
| GuestInfo Guest Id        | other4xLinux64Guest                           |
+---------------------------------------------------------------------------+
| GuestInfo Guest Full Name | Other 4.x Linux (64-bit)                      |
+---------------------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                                    |
+---------------------------------------------------------------------------+
| GuestInfo Detailed Data   | bitness='64'                                  |
|                           | distroName='AlmaLinux'                        |
|                           | distroVersion='8.4'                           |
|                           | familyName='Linux'                            |
|                           | kernelVersion='4.18.0-305.el8.x86_64'         |
|                           | prettyName='AlmaLinux 8.4 (Electric Cheetah)' |
+---------------------------------------------------------------------------+


Test Results (Total: 5, Failed: 1, No Run: 0, Elapsed Time: 01:15:46):
+----------------------------------------------------------+
| Name                              |   Status | Exec Time |
+----------------------------------------------------------+
| deploy_vm_efi_paravirtual_vmxnet3 |   Passed | 00:26:19  |
| gosc_perl_dhcp                    |   Passed | 00:05:27  |
| gosc_perl_staticip                | * Failed | 00:11:17  |
| gosc_cloudinit_dhcp               |   Passed | 00:05:02  |
| gosc_cloudinit_staticip           |   Passed | 00:04:59  |
+----------------------------------------------------------+

Failed log:
2021-09-14 11:20:18,014 | TASK [Assert perl guest customization checks all PASS] *****
task path: /root/ansible-vsphere-gos-validation/linux/guest_customization/linux_gosc_verify.yml:41
failed: [localhost] => (item={'key': 'gosc_log_success', 'value': False}) => {
    "ansible_loop_var": "item",
    "assertion": false,
    "changed": false,
    "evaluated_to": false,
    "item": {
        "key": "gosc_log_success",
        "value": false
    },
    "msg": [
        "Found 'uncaught signal 15' in GOSC log, which should not exist in GOSC log"
    ]
}

```

**Comment by keirazhang** [comment]
> Perl GOSC failed with Photon OS 3.0, need to double check root cause.
> 

Photon OS 3.0 Perl GOSC need to upgrade open-vm-tools to 11.3.0.

```
VM information:
+------------------------------------------------------------------------------+
| Name                      | sle-15-sp3                                       |
+------------------------------------------------------------------------------+
| IP                        | 10.187.128.87                                    |
+------------------------------------------------------------------------------+
| Guest OS Type             | SLES 15.3 x86_64                                 |
+------------------------------------------------------------------------------+
| VMTools Version           | 11.2.5.26209 (build-17337674)                    |
+------------------------------------------------------------------------------+
| CloudInit Version         | 20.2                                             |
+------------------------------------------------------------------------------+
| Config Guest Id           | sles15_64Guest                                   |
+------------------------------------------------------------------------------+
| Hardware Version          | vmx-19                                           |
+------------------------------------------------------------------------------+
| GuestInfo Guest Id        | sles15_64Guest                                   |
+------------------------------------------------------------------------------+
| GuestInfo Guest Full Name | SUSE Linux Enterprise 15 (64-bit)                |
+------------------------------------------------------------------------------+
| GuestInfo Guest Family    | linuxGuest                                       |
+------------------------------------------------------------------------------+
| GuestInfo Detailed Data   | bitness='64'                                     |
|                           | distroName='SLES'                                |
|                           | distroVersion='15.3'                             |
|                           | familyName='Linux'                               |
|                           | kernelVersion='5.3.18-57-default'                |
|                           | prettyName='SUSE Linux Enterprise Server 15 SP3' |
+------------------------------------------------------------------------------+


Test Results (Total: 4, Failed: 0, No Run: 0, Elapsed Time: 00:33:36):
+----------------------------------------------+
| Name                    | Status | Exec Time |
+----------------------------------------------+
| gosc_perl_dhcp          | Passed | 00:06:16  |
| gosc_perl_staticip      | Passed | 00:05:08  |
| gosc_cloudinit_dhcp     | Passed | 00:09:59  |
| gosc_cloudinit_staticip | Passed | 00:09:36  |
+----------------------------------------------+
```

File: linux/guest_customization/linux_gosc_start.yml [hunk 2]
```diff
@@ -41,9 +41,42 @@
     type: "{{ ip_assignment }}"
     timeout: 900
 
-# Wait for GOSC completed
 - include_tasks: ../../common/vm_wait_gosc_completed.yml
 
+# Wait for GOS is rebooted by checking tools status
+- block:
```
Comment (Tomorrow9) [inline linux/guest_customization/linux_gosc_start.yml:47]: I think this is not a good way to say VM is reboot or not just according to the VMware tools status got at that time. What if VM reboots and tools status is running at the time getting tools status, what if there is something wrong with tools service after GOSC and it is not running at the time getting tools status. It's hard to say VM is rebooting, rebooted, or not.

If check other configurations is enough here for now, we can consider to add this check later when there is better solution found.

**Comment by keirazhang** [comment]
Latest regression testing result:
Photon OS 4.0, RHEL/AlmaLinux/RockyLinux/OracleLinux 8.4, SLES 15 SP3 passed GOSC testing. Only Ubuntu 21.04 failed at  gosc_cloudinit_dhcp due to the known issue. 
```
- '[SUCCESS] https://gosv.svc.eng.vmware.com/job/Ansible_Regression_Photon_4.0_OVA/344'
  - 'GUEST OS: VMware Photon OS 4.0 x86_64'
  - 'Results: PASS (5/5) in 00:54:11.'
  - ' * deploy_vmwarephoton_ova...............................................Passed'
  - ' * gosc_perl_dhcp........................................................Passed'
  - ' * gosc_perl_staticip....................................................Passed'
  - ' * gosc_cloudinit_dhcp...................................................Passed'
  - ' * gosc_cloudinit_staticip...............................................Passed'
- '[SUCCESS] https://gosv.svc.eng.vmware.com/job/Ansible_Regression_AlmaLinux_8.4/17'
  - 'GUEST OS: AlmaLinux 8.4 x86_64'
  - 'Results: PASS (5/5) in 01:08:50.'
  - ' * deploy_vm_efi_paravirtual_vmxnet3.....................................Passed'
  - ' * gosc_perl_dhcp........................................................Passed'
  - ' * gosc_perl_staticip....................................................Passed'
  - ' * gosc_cloudinit_dhcp...................................................Passed'
  - ' * gosc_cloudinit_staticip...............................................Passed'
- '[SUCCESS] https://gosv.svc.eng.vmware.com/job/Ansible_Regression_RHEL_8.4/350'
  - 'GUEST OS: RedHat 8.4 x86_64'
  - 'Results: PASS (5/5) in 01:11:23.'
  - ' * deploy_vm_efi_paravirtual_vmxnet3.....................................Passed'
  - ' * gosc_perl_dhcp........................................................Passed'
  - ' * gosc_perl_staticip....................................................Passed'
  - ' * gosc_cloudinit_dhcp...................................................Passed'
  - ' * gosc_cloudinit_staticip...............................................Passed'
- '[SUCCESS] https://gosv.svc.eng.vmware.com/job/Ansible_Regression_RHEL_7.9/39'
  - 'GUEST OS: RedHat 7.9 x86_64'
  - 'Results: PASS (5/5) in 01:15:57.'
  - ' * deploy_vm_efi_paravirtual_vmxnet3.....................................Passed'
  - ' * gosc_perl_dhcp........................................................Passed'
  - ' * gosc_perl_staticip....................................................Passed'
  - ' * gosc_cloudinit_dhcp...................................................Passed'
  - ' * gosc_cloudinit_staticip...............................................Passed'
- '[SUCCESS] https://gosv.svc.eng.vmware.com/job/Ansible_Regression_RockyLinux_8.4/31'
  - 'GUEST OS: Rocky 8.4 x86_64'
  - 'Results: PASS (5/5) in 01:25:04.'
  - ' * deploy_vm_efi_paravirtual_vmxnet3.....................................Passed'
  - ' * gosc_perl_dhcp........................................................Passed'
  - ' * gosc_perl_staticip....................................................Passed'
  - ' * gosc_cloudinit_dhcp...................................................Passed'
  - ' * gosc_cloudinit_staticip...............................................Passed'
- '[SUCCESS] https://gosv.svc.eng.vmware.com/job/Ansible_Regression_OracleLinux_8.4/22'
  - 'GUEST OS: OracleLinux 8.4 x86_64'
  - 'Results: PASS (5/5) in 01:23:59.'
  - ' * deploy_vm_efi_paravirtual_vmxnet3.....................................Passed'
  - ' * gosc_perl_dhcp........................................................Passed'
  - ' * gosc_perl_staticip....................................................Passed'
  - ' * gosc_cloudinit_dhcp...................................................Passed'
  - ' * gosc_cloudinit_staticip...............................................Passed'
customized_test_group:
- '[FAILURE: test_issue] https://gosv.svc.eng.vmware.com/job/Ansible_Regression_SLES_15_SP3/383'
  - 'GUEST OS: SLES 15.3 x86_64'
  - 'Results: PASS (5/5) in 01:33:06.'
  - ' * deploy_vm_efi_paravirtual_vmxnet3.....................................Passed'
  - ' * gosc_perl_dhcp........................................................Passed'
  - ' * gosc_perl_staticip....................................................Passed'
  - ' * gosc_cloudinit_dhcp...................................................Passed'
  - ' * gosc_cloudinit_staticip...............................................Passed'
 ```

```
- '[FAILURE: test_issue] https://gosv.svc.eng.vmware.com/job/Ansible_Regression_Ubuntu_21.04_OVA/29'
  - 'GUEST OS: Ubuntu 21.04 x86_64'
  - 'Results: PASS (4/5), FAIL (1/5) in 01:14:29.'
  - ' * deploy_ubuntu_ova.....................................................Passed'
  - ' * gosc_perl_dhcp........................................................Passed'
  - ' * gosc_perl_staticip....................................................Passed'
  - ' * gosc_cloudinit_dhcp...................................................Failed'
  - ' * gosc_cloudinit_staticip...............................................Passed'
 ```

**Review by Tomorrow9 — APPROVED** [review]
LGTM, thanks.

**Pull request merged**